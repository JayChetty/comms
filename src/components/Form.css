form{
  display: flex;
  flex-direction:column;
  align-items: center;
}
.Form-listitem{
  /*display:flex;
  flex-direction:row;*/
  /*justify-content: center;*/
}
