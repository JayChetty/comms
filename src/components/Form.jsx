import React, { Component } from 'react';
import './Form.css'
import TextField from 'material-ui/TextField';
import {Card, CardActions, CardHeader, CardMedia, CardTitle, CardText} from 'material-ui/Card';
import Chip from 'material-ui/Chip';
import Avatar from 'material-ui/Avatar';
import FontIcon from 'material-ui/FontIcon';
import {greenA200, amber500, blueGrey100, brown400} from 'material-ui/styles/colors';
// import logo from './logo.svg';

// import { connect } from 'react-redux'
// import {
//   Link,
// } from 'react-router-dom'


export default function Form( { form, submission, onFormChange, isCurrentUser, member, memberId, event, score, position} ) {
  if(!submission && isCurrentUser){//setup form from default
    Object.keys(form.default).forEach((formKey)=>{
      onFormChange(form.default[formKey], formKey)
    })
  }

  const dataSource = submission || form.default

  const keysInOrder = ["conservatives", "labour", "snp", "liberals", "other"]
  // // const events = props.events
  const formListItems = keysInOrder.map((formKey, index)=>{

    let label = form.labels[formKey]
    if(event.result){
      const actual = event.result[formKey]
      const error = dataSource[formKey] - actual
      const sign = error >= 0 ? "+" : "-"
      label = `${label} ${actual}( ${sign}${Math.abs(error)})`
    }
    return(
      // <div className="Form-listitem">
      //   <label className="Form-label" key={formKey} for={formKey}> {formKey}</label>
      //   { input }
      // </div>
      <div className="Form-listitem" key={formKey}>
        <TextField
           onChange={ (event)=>onFormChange(event.target.value, formKey) }
           name={formKey}
           hintText={formKey}
           floatingLabelText={label}
           type="number"
           value={dataSource[formKey]}
           autoFocus={index===0}
           disabled={!isCurrentUser || event.status != "open"}
        />
      </div>
    )
  })

  const sum = Object.values( dataSource ).reduce( (acc,curr)=> acc + Number(curr), 0 )
  const remaining = 650 - sum

  let infoBox = <Avatar size={32} backgroundColor={greenA200}>{ sum }</Avatar>

  if(!event.result && remaining != 0){
    const adviceText = remaining > 0 ? 'Add' : 'Remove';
    infoBox = (
    <Chip>
      <Avatar size={32}>{ sum }</Avatar>
      { `${adviceText} ${ Math.abs(remaining) } seats`}
    </Chip>)
  }

  if(event.result){
    let color = null
    switch (position) {
      case 1:
        color = amber500
        break;
      case 2:
        color = blueGrey100
        break;
      case 3:
        color = brown400
        break;
      default:
        color = null
    }
    infoBox = (
      <Chip>
        <Avatar size={32} backgroundColor={color}>{ position }</Avatar>
        { `${score} wrong seats` }
      </Chip>
    )
  }

  return (
    <Card  zDepth={4} key={memberId} style={{marginBottom:'10px'}}>
      <CardHeader
        title={member.displayName}
      />
      <CardText>
        {infoBox}
        <form>
          { formListItems }
        </form>
      </CardText>
    </Card>
  );
}
