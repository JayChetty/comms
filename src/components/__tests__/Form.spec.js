import React from 'react';
import Form from '../Form';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
// it('renders correctly', () => {
// })
const form = {
  default:{a:1, b:2},
  labels: {a:"A", b:"B"},
  name: "Test",
  status: "open"
}

const member={
  displayName:"Jay"
}

it('renders correctly', () => {
  const tree = renderer.create(
    <MuiThemeProvider>
      <Form
        form={form}
        event={{}}
        member={member}
      />
    </MuiThemeProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
