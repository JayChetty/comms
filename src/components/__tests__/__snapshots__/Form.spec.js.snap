// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "WebkitTapHighlightColor": "rgba(0,0,0,0)",
      "backgroundColor": "#ffffff",
      "borderRadius": 2,
      "boxShadow": "0 10px 30px rgba(0, 0, 0, 0.19),
             0 6px 10px rgba(0, 0, 0, 0.23)",
      "boxSizing": "border-box",
      "color": "rgba(0, 0, 0, 0.87)",
      "fontFamily": "Roboto, sans-serif",
      "marginBottom": "10px",
      "muiPrepared": true,
      "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
      "width": "80%",
      "zIndex": 1,
    }
  }
>
  <div
    style={
      Object {
        "paddingBottom": 8,
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "fontWeight": 500,
          "muiPrepared": true,
          "padding": 16,
          "position": "relative",
          "whiteSpace": "nowrap",
        }
      }
    >
      <div
        style={
          Object {
            "display": "inline-block",
            "muiPrepared": true,
            "paddingRight": "90px",
            "verticalAlign": "top",
            "whiteSpace": "normal",
          }
        }
      >
        <span
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "display": "block",
              "fontSize": 15,
              "muiPrepared": true,
            }
          }
        >
          Jay
        </span>
        <span
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.54)",
              "display": "block",
              "fontSize": 14,
              "muiPrepared": true,
            }
          }
        />
      </div>
    </div>
    <div
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontSize": 14,
          "muiPrepared": true,
          "padding": 16,
        }
      }
    >
      <div
        className={undefined}
        disabled={undefined}
        href={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        onTouchTap={[Function]}
        style={
          Object {
            "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
            "backgroundColor": "rgb(224, 224, 224)",
            "border": 10,
            "borderRadius": 16,
            "boxShadow": null,
            "boxSizing": "border-box",
            "cursor": "default",
            "display": "flex",
            "fontFamily": "Roboto, sans-serif",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "margin": 0,
            "muiPrepared": true,
            "outline": "none",
            "padding": 0,
            "position": "relative",
            "textDecoration": "none",
            "verticalAlign": null,
            "whiteSpace": "nowrap",
            "width": "fit-content",
            "zIndex": 1,
          }
        }
        tabIndex={0}
      >
        <div
          className={undefined}
          size={32}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgb(188, 188, 188)",
              "borderRadius": "50%",
              "color": "#ffffff",
              "display": "inline-flex",
              "fontSize": 16,
              "height": 32,
              "justifyContent": "center",
              "marginRight": -4,
              "muiPrepared": true,
              "userSelect": "none",
              "width": 32,
            }
          }
        >
          3
        </div>
        <span
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontSize": 14,
              "fontWeight": 400,
              "lineHeight": "32px",
              "muiPrepared": true,
              "paddingLeft": 12,
              "paddingRight": 12,
              "userSelect": "none",
              "whiteSpace": "nowrap",
            }
          }
        >
          Add 647 seats
        </span>
      </div>
      <form>
        <div
          className="Form-listitem"
        >
          <div
            className={undefined}
            style={
              Object {
                "backgroundColor": "transparent",
                "cursor": "not-allowed",
                "display": "inline-block",
                "fontFamily": "Roboto, sans-serif",
                "fontSize": 16,
                "height": 72,
                "lineHeight": "24px",
                "muiPrepared": true,
                "position": "relative",
                "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": 256,
              }
            }
          >
            <label
              className={undefined}
              htmlFor="a-a-A-11723"
              onTouchTap={undefined}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.3)",
                  "lineHeight": "22px",
                  "muiPrepared": true,
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": 38,
                  "transform": "scale(0.75) translate(0, -28px)",
                  "transformOrigin": "left top",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "userSelect": "none",
                  "zIndex": 1,
                }
              }
            >
              A
            </label>
            <div
              style={
                Object {
                  "bottom": 12,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "muiPrepared": true,
                  "opacity": 0,
                  "position": "absolute",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                }
              }
            >
              a
            </div>
            <input
              autoFocus={true}
              disabled={true}
              id="a-a-A-11723"
              name="a"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              style={
                Object {
                  "WebkitTapHighlightColor": "rgba(0,0,0,0)",
                  "appearance": "textfield",
                  "backgroundColor": "rgba(0,0,0,0)",
                  "border": "none",
                  "boxSizing": "border-box",
                  "color": "rgba(0, 0, 0, 0.3)",
                  "cursor": "inherit",
                  "font": "inherit",
                  "height": "100%",
                  "marginTop": 14,
                  "muiPrepared": true,
                  "outline": "none",
                  "padding": 0,
                  "position": "relative",
                  "width": "100%",
                }
              }
              type="number"
              value={1}
            />
            <div>
              <hr
                aria-hidden="true"
                style={
                  Object {
                    "borderBottom": "dotted 2px",
                    "borderColor": "rgba(0, 0, 0, 0.3)",
                    "borderLeft": "none",
                    "borderRight": "none",
                    "borderTop": "none",
                    "bottom": 8,
                    "boxSizing": "content-box",
                    "margin": 0,
                    "muiPrepared": true,
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              />
              <hr
                aria-hidden="true"
                style={
                  Object {
                    "borderBottom": "solid 2px",
                    "borderColor": "#00bcd4",
                    "borderLeft": "none",
                    "borderRight": "none",
                    "borderTop": "none",
                    "bottom": 8,
                    "boxSizing": "content-box",
                    "margin": 0,
                    "muiPrepared": true,
                    "position": "absolute",
                    "transform": "scaleX(0)",
                    "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                    "width": "100%",
                  }
                }
              />
            </div>
          </div>
        </div>
        <div
          className="Form-listitem"
        >
          <div
            className={undefined}
            style={
              Object {
                "backgroundColor": "transparent",
                "cursor": "not-allowed",
                "display": "inline-block",
                "fontFamily": "Roboto, sans-serif",
                "fontSize": 16,
                "height": 72,
                "lineHeight": "24px",
                "muiPrepared": true,
                "position": "relative",
                "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": 256,
              }
            }
          >
            <label
              className={undefined}
              htmlFor="b-b-B-33594"
              onTouchTap={undefined}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.3)",
                  "lineHeight": "22px",
                  "muiPrepared": true,
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": 38,
                  "transform": "scale(0.75) translate(0, -28px)",
                  "transformOrigin": "left top",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "userSelect": "none",
                  "zIndex": 1,
                }
              }
            >
              B
            </label>
            <div
              style={
                Object {
                  "bottom": 12,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "muiPrepared": true,
                  "opacity": 0,
                  "position": "absolute",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                }
              }
            >
              b
            </div>
            <input
              autoFocus={false}
              disabled={true}
              id="b-b-B-33594"
              name="b"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              style={
                Object {
                  "WebkitTapHighlightColor": "rgba(0,0,0,0)",
                  "appearance": "textfield",
                  "backgroundColor": "rgba(0,0,0,0)",
                  "border": "none",
                  "boxSizing": "border-box",
                  "color": "rgba(0, 0, 0, 0.3)",
                  "cursor": "inherit",
                  "font": "inherit",
                  "height": "100%",
                  "marginTop": 14,
                  "muiPrepared": true,
                  "outline": "none",
                  "padding": 0,
                  "position": "relative",
                  "width": "100%",
                }
              }
              type="number"
              value={2}
            />
            <div>
              <hr
                aria-hidden="true"
                style={
                  Object {
                    "borderBottom": "dotted 2px",
                    "borderColor": "rgba(0, 0, 0, 0.3)",
                    "borderLeft": "none",
                    "borderRight": "none",
                    "borderTop": "none",
                    "bottom": 8,
                    "boxSizing": "content-box",
                    "margin": 0,
                    "muiPrepared": true,
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              />
              <hr
                aria-hidden="true"
                style={
                  Object {
                    "borderBottom": "solid 2px",
                    "borderColor": "#00bcd4",
                    "borderLeft": "none",
                    "borderRight": "none",
                    "borderTop": "none",
                    "bottom": 8,
                    "boxSizing": "content-box",
                    "margin": 0,
                    "muiPrepared": true,
                    "position": "absolute",
                    "transform": "scaleX(0)",
                    "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                    "width": "100%",
                  }
                }
              />
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
